### PR TITLE
tpm2_policysecret: New tool to enable object secret based authentication to policies

### PR DESCRIPTION
**Pull request by idesai**

### CODE REVIEW

File: lib/tpm2_policy.h [hunk 1]
```diff
@@ -99,6 +99,25 @@ bool tpm2_policy_build_policyauthorize(
 bool tpm2_policy_build_policyor(TSS2_SYS_CONTEXT *sapi_context,
     tpm2_session *policy_session, TPML_DIGEST policy_list);
 
+/**
+ * Enables secret (password/ hmac) based authorization to a policy.
```
Comment (williamcroberts) [inline lib/tpm2_policy.h]: weird space on `password/ hmac`

File: man/tpm2_policysecret.1.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+% tpm2_policysecret(1) tpm2-tools | General Commands Manual
+%
+% OCTOBER 2018
+
+# NAME
+
+**tpm2_policysecret**(1) - Enables secret(password/ hmac) based authorization to a policy.
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: space on `password/ hmac`

File: man/tpm2_policysecret.1.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+# SYNOPSIS
+
+**tpm2_policysecret** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_policysecret**(1) Enables secret (password/ hmac) based authorization to a policy. The secret is the authentication value of any TPM object (NV/ Hierarchy/ Loaded/ Persistent Handles).
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: same space issue, do you need to escape the `/` perhaps?

File: man/tpm2_policysecret.1.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# OPTIONS
+
+  * **-c**, **--context**=_OBJECT_CONTEXT_:
+
+    A context object specifier of a transient/ permanent/ persistent object. Either a file path of a object context blob or a loaded/ persistent/ permanent handle id. See section "Context Object Format".
+
+  * **-p**, **--auth-value**=_AUTH\_VALUE_:
+
+    The authentication value of the object whose secret will be referred in the policy. Authentication values should follow the "authorization formatting standards", see section "Authorization Formatting".
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: Is that true? Won't the only secret be a string? What about binary data?

Doesn't the auth values include things like:
"foo"
"hex:aa11bb"
"session:session.ctx"

session would be unsupported.

Comment (williamcroberts) [reply man/tpm2_policysecret.1.md]: Make this an argument, and if it's missing read from stdin.

File: man/tpm2_policysecret.1.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+[common tcti options](common/tcti.md)
+
+[supported hash algorithms](common/hash.md)
+
+[algorithm specifiers](common/alg.md)
+
+# EXAMPLES
+
+Choose a TPM handle like TPM_RH_OWNER. Supply the 32 bit handle # of the object to the **tpm2_policysecret** tool along with the authentication value for the object in a *trial* session. Use the resultant policy digest as 
+input authentication policy when creating a sealing object. In a *real* policy session supply the auth object's authentication data along with object's context/ handle id, if the secret provided is correct, unsealing should succeed.
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: The first few sentences should explain what your accomplishing not how you're doing it. Please re-write this with the opening statement being what you want to accomplish, followed by how you will accomplish it.

File: man/tpm2_policysecret.1.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+## Generate a policy that binds to the secret of the auth object:
+* tpm2_startauthsession -S session.ctx
+* tpm2_policysecret -S session.ctx -c 0x40000001 -o secret.policy
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: This nv index comes out of nowhere. You should include creating the nvindex so it's not a magic number.

File: test/integration/tests/tcti/abrmd/policysecret.sh [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+echo "40000001" | xxd -r -p > $object_name
+cat $policy_init $policysecret_cc $object_name > $concatenated
+openssl dgst -binary -sha256 $concatenated |\
+openssl dgst -binary -sha256 > $test_vector
+
+tpm2_startauthsession -S $session_ctx
+tpm2_policysecret -o $o_policy_digest -S $session_ctx -c 0x40000001
+tpm2_flushcontext -S $session_ctx 
+
+diff $test_vector $o_policy_digest
```
Comment (williamcroberts) [inline test/integration/tests/tcti/abrmd/policysecret.sh]: That's an interesting way to test this, can you also add a test actually authenticating to an object like in your example code.

File: tools/tpm2_policysecret.c [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    if (!ctx.flags.c) {
+        LOG_ERR("Must specify -c handle-id/ context file path.");
+        return false;
+    }
+
+    /*
+     *It is possible to have the authorization secret of the auth object NULL
+     */
```
Comment (williamcroberts) [inline tools/tpm2_policysecret.c]: is this really going to be a common use case?

If you stall waiting on stdin, folks can use `ctrl + d` to indicate EOF to put in empty.

Comment (williamcroberts) [reply tools/tpm2_policysecret.c]: They could also pass the empty string.

Also, you're argument handling should be arg count >= 0. If 0 args read stdin, else for each argument issue a policy password event.

**Comment by idesai** [comment]
Depends on #1230

File: tools/tpm2_policysecret.c (hunk 8 above)
Comment (idesai) [reply tools/tpm2_policysecret.c]: @williamcroberts added fmemopen to files handling routing to read from stdin. LMK. Thanks.

File: tools/tpm2_policysecret.c [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+    TPM2B_DIGEST policy_digest = TPM2B_EMPTY_INIT;
+
+    result = tpm2_policy_get_digest(sapi_context, s, &policy_digest);
+    if (!result) {
+        LOG_ERR("Could not build tpm policy");
+        goto out;
+    }
+
+    tpm2_util_hexdump(policy_digest.buffer, policy_digest.size);
```
Comment (williamcroberts) [inline tools/tpm2_policysecret.c]: print a new line using the tpm2_output routine.

File: tools/tpm2_policysecret.c [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file", required_argument, NULL, 'o' },
+        { "session",     required_argument, NULL, 'S' },
+        { "context",     required_argument, NULL, 'c' },
+        { "auth-value",  required_argument, NULL, 'p' },
```
Comment (williamcroberts) [inline tools/tpm2_policysecret.c]: I though this was supposed to be an argument?

File: tools/tpm2_policysecret.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        ctx.out_policy_dgst_path = value;
+        break;
+    case 'S':
+        ctx.session_path = value;
+        break;
+    case 'c':
+        ctx.context_arg = value;
+        ctx.flags.c = 1;
+        break;
+    case 'p':
```
[the remark this answers is not in the document]
Comment (williamcroberts) [reply tools/tpm2_policysecret.c]: Still need to handle

File: tools/tpm2_policysecret.c [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        break;
+    }
+
+    return result;
+}
+
+bool on_arg (int argc, char **argv) {
+
+    if (argc > 1) {
+        LOG_ERR("Specify stdin with '-' ");
```
Comment (williamcroberts) [inline tools/tpm2_policysecret.c]: Huh?

```C
LOG_ERR("Specify a single auth value");
```

File: tools/tpm2_policysecret.c [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        return true;
+    }
+
+    bool result;
+    if (!strcmp(argv[0], "-")) {
+        //auth data from stdin
+        result = files_load_bytes_from_buffer_or_file_or_stdin(NULL,NULL,
+                &ctx.auth_str_size, ctx.session_data.hmac.buffer);
+        if (!result) {
+            LOG_ERR("Failed stdin input for auth.");
```
Comment (williamcroberts) [inline tools/tpm2_policysecret.c]: I think files_load_bytes_from_buffer_or_file_or_stdin() logs all errors. So just drop this or fix up the sentence:
```C
LOG_ERR("Failed to read auth value from stdin");
```

File: tools/tpm2_policysecret.c (hunk 9 above)
Comment (williamcroberts) [reply tools/tpm2_policysecret.c]: Still open

File: scripts/utils/analyze.py [hunk 14]
```diff
@@ -134,10 +134,10 @@ def __init__(self, tools, full=False):
             },
             {
                 "gname": "ea",
-                "tools-in-group": ["tpm2_policypcr", "tpm2_createpolicy", "tpm2_policyauthorize", "tpm2_policyor", "tpm2_policypassword"],
+                "tools-in-group": ["tpm2_policypcr", "tpm2_createpolicy", "tpm2_policyauthorize", "tpm2_policyor", "tpm2_policypassword", "tpm2_policysecret"],
                 "tools": [],
                 "conflict": None,
-                "ignore": set(['S', 'session', 'q', 'qualifier', 'n', 'name', 't', 'ticket', 'L', 'policy-list', 'o', 'policy-file'])
+                "ignore": set(['S', 'session', 'q', 'qualifier', 'n', 'name', 't', 'ticket', 'L', 'policy-list', 'o', 'policy-file', 'c', 'context', 'p', 'auth-value'])
```
Comment (williamcroberts) [inline scripts/utils/analyze.py]: You don't need the p in here anymore

File: man/tpm2_policysecret.1.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+# SYNOPSIS
+
+**tpm2_policysecret** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_policysecret**(1) Enables secret (password/hmac) based authorization to a
+ policy. The secret is the authentication value of any TPM object (NV/Hierarchy/
+Loaded/Persistent Handles).
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: missing space indentation.

File: man/tpm2_policysecret.1.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+[common tcti options](common/tcti.md)
+
+[supported hash algorithms](common/hash.md)
+
+[algorithm specifiers](common/alg.md)
+
+# EXAMPLES
+
+Tying the auth value of a sealing object to the owner hierarchy password: Using
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: replace `:` with `.`

File: man/tpm2_policysecret.1.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+ **tpm2_policysecret** in a trial session, create policy that can only be
+ satisfied if owner hierarchy auth is supplied. In a real policy session supply
+ the owner hierarchy auth value. If the owner auth value provided is correct,
+ unsealing should succeed.
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: no space in beginning of lines here, for lines 48-51

File: man/tpm2_policysecret.1.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+% OCTOBER 2018
+
+# NAME
+
+**tpm2_policysecret**(1) - Enables secret(password/hmac) based authorization to
+a policy.
+
+# SYNOPSIS
+
+**tpm2_policysecret** [*OPTIONS*]
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: Need to mention the argument `_AUTH_VALUE_`

```
**tpm2_policysecret** [*OPTIONS*] _AUTH\_VALUE_
```

File: man/tpm2_policysecret.1.md (hunk 15 above)
Comment (williamcroberts) [reply man/tpm2_policysecret.1.md]: Also describe the argument. As an argument, it takes the auth value of the associated TPM object, a single dash `-` can be used to read the auth value from stdin. The argument follows the "formatting standards for passwords ... (go find that verbatim text and drop it in here with the see that reference stuff)

File: man/tpm2_policysecret.1.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+  * **-o**, **--policy-file**=_POLICY\_FILE_:
+
+    File to save the policy digest.
+
+[common options](common/options.md)
+
+[common tcti options](common/tcti.md)
+
+[supported hash algorithms](common/hash.md)
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: It has hash algorithms? Remove

File: man/tpm2_policysecret.1.md [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+[algorithm specifiers](common/alg.md)
```
Comment (williamcroberts) [inline man/tpm2_policysecret.1.md]: It takes algorithms, remove

**Comment by williamcroberts** [comment]
@idesai please resolve the merge conflict via git pull --rebase origin master and push again

**Pull request merged**